AK/SipHash.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/MemoryStream.h>
+#include <AK/SipHash.h>
+#include <AK/Span.h>
+#include <AK/UFixedBigInt.h>
+
+#ifdef KERNEL
+#    include <Kernel/Security/Random.h>
+#else
+#    include <AK/Random.h>
+#endif
+
+namespace AK {
+
+#define ROTL(x, b) (u64)(((x) << (b)) | ((x) >> (64 - (b))))
+
+#define SIPROUND           \
+    do {                   \
+        v0 += v1;          \
+        v1 = ROTL(v1, 13); \
+        v1 ^= v0;          \
+        v0 = ROTL(v0, 32); \
+        v2 += v3;          \
+        v3 = ROTL(v3, 16); \
+        v3 ^= v2;          \
+        v0 += v3;          \
+        v3 = ROTL(v3, 21); \
+        v3 ^= v0;          \
+        v2 += v1;          \
+        v1 = ROTL(v1, 17); \
+        v1 ^= v2;          \
+        v2 = ROTL(v2, 32); \
+    } while (0)
+
+template<size_t message_block_rounds, size_t finalization_rounds>
+static void do_siphash(ReadonlyBytes input, u128 key, Bytes output)
```

Comment:
Looking further up now, your `output` is either the underlying bytes of a u64...or a u64, so when is `output.size() == 16`?

Reply:
The code is general over outputs of 8 or 16 bytes, though we can't use the 16-byte output conveniently at the moment since u128 cannot be included from the SipHash.h header due to include cycles from Format.h. I still want this general code to be available so we can implement that later, even if we need to use a raw Bytes object instead of a convenient u128.

Reply:
err...what does the header have to do with this .cpp file?
regardless, at least make a note of that.

We're already chopping bits off the 64-bit output, what's the point of having 128 bits when you're just gonna take 32 of them?